### DISCUSSION TITLE
libp2p_kad::handler No older substream waiting to be reused. Dropping new substream.

### DISCUSSION THREAD

**Discussion by lynic** [libp2p/rust-libp2p#2957] [Q&A] [answered]
I try to setup a kad network, I enabled kad,relay,ping,identify,gossipsub in my code, and setup one as server with public access.

Then I spawn 10~20 clients to connect to this server, client enables kad,ping,identify,gossipsub, and use kad find_closest_peers to update DHT.

After some time, the server logs shows many message like below.
```
[2022-09-28T16:32:45Z WARN  libp2p_kad::handler] New inbound substream to PeerId("12D3KooWLMtdcMzvCQH5NseLaBbBqETrRLQuQCVSjsGsb4qzutD8") exceeds inbound substream limit. No older substream waiting to be reused. Dropping new substream.
```

Would like to understand what does this message mean? and how to remediate it?

**Comment by thomaseizinger** [comment, marked as answer]
To protect against DoS attacks, the kademlia implementation limits the amount of inbound streams that a single remote peer can open on one connection. This limit is currently set to 32 and not configurable.

I don't know kademlia in depth (@mxinden does) but 32 streams PER connection sounds like a fair amount to me. It essentially means that a single peer is currently keeping us busy with 32 _concurrent_ requests. Once a request is serviced, the same substream can be re-used again. That is the reuse the message is talking about.

Essentially, this warning means that your node is under high load and (temporarily) refusing to service even more requests.

**Reply by lynic** [reply]
32 substreams from one peers sounds fine, just not sure why in my env, there're  so many streams from one remote peer. ping/identify will regular send data, another one would be gossipsub, but I didn't use gossipsub to pub msg yet.  Is it possible that kad use that many streams?

**Reply by thomaseizinger** [reply]
Yeah I think that is indeed possible. Most importantly, `kad` has some persistent streams I think whereas most other protocols will close a stream after a message exchange (i.e. identify & ping).

**Reply by lynic** [reply]
Thanks! Do you know how to get streams info for a remote peer?

**Reply by thomaseizinger** [reply]
You mean the number of open streams? I am [working](https://github.com/libp2p/rust-libp2p/pull/2878) on exposing that. Once that PR is merged, we can introduce metrics that allow you to access this information.

**Reply by mxinden** [reply]
Agreed with everything by @thomaseizinger above. Thanks for helping out here.

While we enforce a limit at the listening end, we do not enforce a limit on the sending end. That is less than ideal and the reason why you see the error message.

I think long term we should enforce a limit on the sending end. Ideally this would include adding a recommendation in https://github.com/libp2p/specs/blob/master/kad-dht/README.md for other implementations.

@lynic let me know in case you would like to work on this. Note that it is not as easy as it might seem at first though.

//CC @dignifiedquire as I expect you will run into this long term as well.

**Reply by lynic** [reply]
Thanks all! I suspect this cause by my incorrect usage of  `gossipsub.add_explicit_peer` , now before calling this func, I will check if this peer is already exist in gossipsub. After this modification, this issue seems disappeared.

**Reply by dignifiedquire** [reply]
Did something change? I realized I have started seeing this error since the last update too now

**Reply by thomaseizinger** [reply]
https://github.com/libp2p/rust-libp2p/pull/2861 might be related but I can't think how that would trigger the above error because we now stop creating substreams once we hit the limit instead of passing them on so if anything, it should have the opposite effect 🤷‍♂️

**Reply by dignifiedquire** [reply]
FYI these happen a lot
<img width="1512" alt="Screen Shot 2022-10-10 at 18 41 06" src="https://user-images.githubusercontent.com/790842/194914870-12fcc147-0d45-491d-a01e-34e739890763.png">

**Reply by dignifiedquire** [reply]
I am not entirely clear how I can do sth on my end to prevent this, I have a lot of queries to run 😅

**Reply by thomaseizinger** [reply]
With https://github.com/libp2p/rust-libp2p/pull/2878, these will go away and we will just not accept any more inbound streams from the remote. Depending on the muxer, this may lead to timeouts on the remote in opening new streams.

**Reply by mxinden** [reply]
> With #2878, these will go away and we will just not accept any more inbound streams from the remote. Depending on the muxer, this may lead to timeouts on the remote in opening new streams.

I am not convinced a global limit is the way to go. See also https://github.com/libp2p/rust-libp2p/pull/2878#discussion_r980327832.


> I am not entirely clear how I can do sth on my end to prevent this, I have a lot of queries to run sweat_smile

In my eyes the best long-term fix is to add a maximum number of concurrent streams to the Kademlia specification. I do think it is reasonable to assume that a node can not make use of a parallelization factor of say 16 to a remote peer. @dignifiedquire would you be willing to drive this?

**Reply by thomaseizinger** [reply]
> > With #2878, these will go away and we will just not accept any more inbound streams from the remote. Depending on the muxer, this may lead to timeouts on the remote in opening new streams.
> 
> I am not convinced a global limit is the way to go. See also [#2878 (comment)](https://github.com/libp2p/rust-libp2p/pull/2878#discussion_r980327832).

The latest version of #2878 allows the `ConnectionHandler` to decide on the max number of inbound streams. Once hit, the muxer on this connection will no longer poll for inbound streams.

For Kademlia, this is now set to the 32 that was enforced previously as part of the event handler in the `ConnectionHandler`. With "this will go away", I was referring to the log message. That one is removed because the `ConnectionHandler` is now guaranteed to not be given more substreams than what it reported in `max_inbound_streams` :)

**Comment by shamil-gadelshin** [comment]
I have a similar issue with just two peers: https://github.com/libp2p/rust-libp2p/issues/3048
